### PR TITLE
MediaBrowser.Api code cleanup

### PR DESCRIPTION
**Pull request by barronpm**
**Changes**

Contains miscellaneous code cleanup throughout MediaBrowser.Api
Most changes are pretty self-explanatory

### CODE REVIEW

File: MediaBrowser.Api/ItemLookupService.cs [hunk 1]
```diff
@@ -305,9 +305,15 @@ private async Task DownloadImage(string providerName, string url, Guid urlHash,
 
             Directory.CreateDirectory(Path.GetDirectoryName(fullCachePath));
             using (var stream = result.Content)
-            using (var filestream = new FileStream(fullCachePath, FileMode.Create, FileAccess.Write, FileShare.Read, IODefaults.FileStreamBufferSize, true))
             {
-                await stream.CopyToAsync(filestream).ConfigureAwait(false);
+                using var fileStream = new FileStream(fullCachePath,
```
Comment (Bond-009) [inline MediaBrowser.Api/ItemLookupService.cs:309]:
```suggestion
                using var fileStream = new FileStream(
                    fullCachePath,
```

File: MediaBrowser.Api/Library/LibraryService.cs [hunk 2]
```diff
@@ -397,22 +388,18 @@ private bool IsMetadataFetcherEnabledByDefault(string name, string type, bool is
             {
                 if (string.Equals(name, "TheMovieDb", StringComparison.OrdinalIgnoreCase))
                 {
+
                     if (string.Equals(type, "Series", StringComparison.OrdinalIgnoreCase))
                     {
                         return true;
                     }
```
Comment (Bond-009) [inline MediaBrowser.Api/Library/LibraryService.cs:395]:
```suggestion
                    }

```

File: MediaBrowser.Api/Library/LibraryService.cs [hunk 3]
```diff
@@ -1086,44 +1063,26 @@ private ThemeMediaResult GetThemeSongs(GetThemeSongs request)
...
-                }
-
-                var parent = item.GetParent();
-                if (parent == null)
-                {
-                    break;
-                }
-                item = parent;
+                item = item.GetParent();
+                themeItems = item.GetThemeSongs();
```
Comment (Bond-009) [inline MediaBrowser.Api/Library/LibraryService.cs:1079]: `GetParent()` can return null, this is a possible `NullRefException`

File: MediaBrowser.Api/Library/LibraryService.cs [hunk 4]
```diff
@@ -1152,44 +1109,27 @@ public ThemeMediaResult GetThemeVideos(GetThemeVideos request)
...
-                }
-
-                var parent = item.GetParent();
-                if (parent == null)
-                {
-                    break;
-                }
-                item = parent;
+                item = item.GetParent();
+                themeItems = item.GetThemeVideos();
```
Comment (Bond-009) [inline MediaBrowser.Api/Library/LibraryService.cs:1125]: Same as above

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 5]
```diff
@@ -137,12 +137,7 @@ private string GetOutputFilePath(StreamState state, EncodingOptions encodingOpti
             var ext = outputFileExtension.ToLowerInvariant();
             var folder = ServerConfigurationManager.GetTranscodePath();
 
-            if (EnableOutputInSubFolder)
-            {
-                return Path.Combine(folder, filename, filename + ext);
-            }
-
-            return Path.Combine(folder, filename + ext);
+            return EnableOutputInSubFolder ? Path.Combine(folder, filename, filename + ext) : Path.Combine(folder, filename + ext);
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs:140]: Split over multiple lines

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 6]
```diff
@@ -389,195 +378,215 @@ private void ParseParams(StreamRequest request)
...
+                        request.DeviceId = val;
+                        break;
+                    case 2:
+                        request.MediaSourceId = val;
+                        break;
+                    case 3:
+                        request.Static = string.Equals("true", val, StringComparison.OrdinalIgnoreCase);
+                        break;
+                    case 4:
                     {
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs:396]: Don't use braces here

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 7]
```diff
@@ -862,14 +866,7 @@ private void ApplyDeviceProfileSettings(StreamState state)
...
 
-                    if (caps != null)
-                    {
-                        state.DeviceProfile = caps.DeviceProfile;
-                    }
-                    else
-                    {
-                        state.DeviceProfile = DlnaManager.GetProfile(headers);
-                    }
+                    state.DeviceProfile = caps != null ? caps.DeviceProfile : DlnaManager.GetProfile(headers);
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs:869]: Again, switch these around

File: MediaBrowser.Api/Playback/Hls/BaseHlsService.cs [hunk 8]
```diff
@@ -247,17 +240,12 @@ protected virtual async Task WaitForMinimumSegmentCount(string playlist, int seg
...
-
-            try
-            {
-                return new FileStream(tmpPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite, IODefaults.FileStreamBufferSize, FileOptions.SequentialScan);
-            }
-            catch (IOException)
-            {
-                return new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite, IODefaults.FileStreamBufferSize, FileOptions.SequentialScan);
-            }
+            return new FileStream(path,
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/Hls/BaseHlsService.cs:243]:
```suggestion
            return new FileStream(
                path,
```

File: MediaBrowser.Api/Playback/MediaInfoService.cs [hunk 9]
```diff
@@ -583,7 +583,7 @@ private async Task<PlaybackInfoResponse> GetPlaybackInfo(Guid id, Guid userId, s
         private long? GetMaxBitrate(long? clientMaxBitrate, User user)
         {
             var maxBitrate = clientMaxBitrate;
-            var remoteClientMaxBitrate = user == null ? 0 : user.Policy.RemoteClientBitrateLimit;
+            var remoteClientMaxBitrate = user?.Policy.RemoteClientBitrateLimit ?? 0;
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/MediaInfoService.cs:586]: Previous was clearer IMO

File: MediaBrowser.Api/SearchService.cs [hunk 10]
```diff
@@ -234,59 +234,50 @@ private SearchHint GetSearchHintResult(SearchHintInfo hintInfo)
...
-            {
-                if (series.Status.HasValue)
+                case IHasSeries hasSeries:
+                    result.Series = hasSeries.SeriesName;
+                    break;
+                case LiveTvProgram program:
+                    result.StartDate = program.StartDate;
+                    break;
+                case Series series:
                 {
```
Comment (Bond-009) [inline MediaBrowser.Api/SearchService.cs:246]: Again, no braces

File: MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs [hunk 11]
```diff
@@ -396,12 +396,10 @@ public string[] GetPersonTypes()
...
-            if (string.IsNullOrEmpty(VideoTypes))
-            {
-                return Array.Empty<VideoType>();
-            }
-
-            return VideoTypes.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries).Select(v => (VideoType)Enum.Parse(typeof(VideoType), v, true)).ToArray();
+            return string.IsNullOrEmpty(VideoTypes)
+                ? Array.Empty<VideoType>()
+                : VideoTypes.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries)
+                    .Select(v => (VideoType)Enum.Parse(typeof(VideoType), v, true)).ToArray();
```
Comment (Bond-009) [inline MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs:402]: Use generic `Enum.Parse`

File: MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs [hunk 12]
```diff
@@ -412,12 +410,10 @@ public ItemFilter[] GetFilters()
...
-            if (string.IsNullOrEmpty(val))
-            {
-                return new ItemFilter[] { };
-            }
-
-            return val.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries).Select(v => (ItemFilter)Enum.Parse(typeof(ItemFilter), v, true)).ToArray();
+            return string.IsNullOrEmpty(val)
+                ? new ItemFilter[] { }
+                : val.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries).
+                    Select(v => (ItemFilter)Enum.Parse(typeof(ItemFilter), v, true)).ToArray();
```
Comment (Bond-009) [inline MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs:416]: Same as above

File: MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs [hunk 13]
```diff
@@ -412,12 +410,10 @@ public ItemFilter[] GetFilters()
...
             var val = Filters;
 
-            if (string.IsNullOrEmpty(val))
-            {
-                return new ItemFilter[] { };
-            }
-
-            return val.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries).Select(v => (ItemFilter)Enum.Parse(typeof(ItemFilter), v, true)).ToArray();
+            return string.IsNullOrEmpty(val)
+                ? new ItemFilter[] { }
```
Comment (Bond-009) [inline MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs:414]: Use `Array.Empty`

File: MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs [hunk 14]
```diff
@@ -428,12 +424,9 @@ public ImageType[] GetImageTypes()
...
 
-            if (string.IsNullOrEmpty(val))
-            {
-                return new ImageType[] { };
-            }
-
-            return val.Split(',').Select(v => (ImageType)Enum.Parse(typeof(ImageType), v, true)).ToArray();
+            return string.IsNullOrEmpty(val)
+                ? new ImageType[] { }
+                : val.Split(',').Select(v => (ImageType)Enum.Parse(typeof(ImageType), v, true)).ToArray();
```
Comment (Bond-009) [inline MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs:429]: Both same as above

File: MediaBrowser.Api/VideosService.cs [hunk 15]
```diff
@@ -138,32 +138,15 @@ public void Post(MergeVersions request)
...
-                            return 1;
-                        }
-
-                        return 0;
-                    })
-                    .ThenByDescending(i =>
-                    {
-                        var stream = i.GetDefaultVideoStream();
+            var primaryVersion = videosWithVersions.FirstOrDefault() ?? items.OrderBy(i
+                    => (i.Video3DFormat.HasValue || i.VideoType != Model.Entities.VideoType.VideoFile) ? 1 : 0)
```
Comment (Bond-009) [inline MediaBrowser.Api/VideosService.cs:142]: Previous was clearer

File: MediaBrowser.Api/Library/LibraryService.cs (hunk 3 above)
Comment (barronpm) [reply MediaBrowser.Api/Library/LibraryService.cs:1079]: This is checked for in the conditions for the while loop

Comment (Bond-009) [reply MediaBrowser.Api/Library/LibraryService.cs:1079]: You already use the result in the next line...

Comment (barronpm) [reply MediaBrowser.Api/Library/LibraryService.cs:1042]: The loop checks if `item.GetParent()` is null, not `item`

File: MediaBrowser.Api/System/SystemService.cs [hunk 16]
```diff
@@ -168,12 +168,8 @@ public Task<object> Get(GetLogFile request)
...
 
             // For older files, assume fully static
-            if (file.LastWriteTimeUtc < DateTime.UtcNow.AddHours(-1))
-            {
-                return ResultFactory.GetStaticFileResult(Request, file.FullName, FileShare.Read);
-            }
-
-            return ResultFactory.GetStaticFileResult(Request, file.FullName, FileShare.ReadWrite);
+            return ResultFactory.GetStaticFileResult(Request, file.FullName,
+                file.LastWriteTimeUtc < DateTime.UtcNow.AddHours(-1) ? FileShare.Read : FileShare.ReadWrite);
```
Comment (mark-monteiro) [inline MediaBrowser.Api/System/SystemService.cs:172]: This is okay but I think using a separate variable would be a bit clearer

```
var fileShare = file.LastWriteTimeUtc < DateTime.UtcNow.AddHours(-1) ? FileShare.Read : FileShare.ReadWrite;
ResultFactory.GetStaticFileResult(Request, file.FullName, fileShare);
```

File: MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs [hunk 17]
```diff
@@ -428,12 +424,9 @@ public ImageType[] GetImageTypes()
...
             var val = ImageTypes;
 
-            if (string.IsNullOrEmpty(val))
-            {
-                return new ImageType[] { };
-            }
-
-            return val.Split(',').Select(v => (ImageType)Enum.Parse(typeof(ImageType), v, true)).ToArray();
+            return string.IsNullOrEmpty(val)
+                ? new ImageType[] { }
```
Comment (mark-monteiro) [inline MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs:428]: Use `Array.Empty<>()` here

File: MediaBrowser.Api/SearchService.cs [hunk 18]
```diff
@@ -323,12 +309,8 @@ private void SetThumbImageInfo(SearchHint hint, BaseItem item)
...
         private void SetBackdropImageInfo(SearchHint hint, BaseItem item)
         {
-            var itemWithImage = item.HasImage(ImageType.Backdrop) ? item : null;
-
-            if (itemWithImage == null)
-            {
-                itemWithImage = GetParentWithImage<BaseItem>(item, ImageType.Backdrop);
-            }
+            var itemWithImage = (item.HasImage(ImageType.Backdrop) ? item : null)
+                                ?? GetParentWithImage<BaseItem>(item, ImageType.Backdrop);
```
Comment (mark-monteiro) [inline MediaBrowser.Api/SearchService.cs:313]: Only indent one tab here

File: MediaBrowser.Api/Library/LibraryService.cs [hunk 19]
```diff
@@ -397,54 +388,22 @@ private bool IsMetadataFetcherEnabledByDefault(string name, string type, bool is
...
-                }
-                else if (string.Equals(name, "TheAudioDB", StringComparison.OrdinalIgnoreCase))
-                {
-                    return true;
-                }
-                else if (string.Equals(name, "MusicBrainz", StringComparison.OrdinalIgnoreCase))
-                {
-                    return true;
+                    return !(string.Equals(type, "Season", StringComparison.OrdinalIgnoreCase)
+                             || string.Equals(type, "Episode", StringComparison.OrdinalIgnoreCase)
```
Comment (mark-monteiro) [inline MediaBrowser.Api/Library/LibraryService.cs:392]: This should just be indented one tab

File: MediaBrowser.Api/Library/LibraryService.cs [hunk 20]
```diff
@@ -397,54 +388,22 @@ private bool IsMetadataFetcherEnabledByDefault(string name, string type, bool is
...
+                             || string.Equals(type, "MusicVideo", StringComparison.OrdinalIgnoreCase));
                 }
 
-                return false;
+                return string.Equals(name, "TheTVDB", StringComparison.OrdinalIgnoreCase)
+                       || string.Equals(name, "TheAudioDB", StringComparison.OrdinalIgnoreCase)
```
Comment (mark-monteiro) [inline MediaBrowser.Api/Library/LibraryService.cs:397]: Same

File: MediaBrowser.Api/Library/LibraryService.cs [hunk 21]
```diff
@@ -397,54 +388,22 @@ private bool IsMetadataFetcherEnabledByDefault(string name, string type, bool is
...
                 .ToArray();
 
-            if (metadataOptions.Length == 0)
-            {
-                return true;
-            }
-
-            return metadataOptions.Any(i => !i.DisabledMetadataFetchers.Contains(name, StringComparer.OrdinalIgnoreCase));
+            return metadataOptions.Length == 0
+                   || metadataOptions.Any(i => !i.DisabledMetadataFetchers.Contains(name, StringComparer.OrdinalIgnoreCase));
```
Comment (mark-monteiro) [inline MediaBrowser.Api/Library/LibraryService.cs:406]: Same

File: MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs [hunk 22]
```diff
@@ -412,12 +410,10 @@ public ItemFilter[] GetFilters()
...
-            if (string.IsNullOrEmpty(val))
-            {
-                return new ItemFilter[] { };
-            }
-
-            return val.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries).Select(v => (ItemFilter)Enum.Parse(typeof(ItemFilter), v, true)).ToArray();
+            return string.IsNullOrEmpty(val)
+                ? Array.Empty<ItemFilter>()
+                : val.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries).
+                    Select(v => Enum.Parse<ItemFilter>(v, true)).ToArray();
```
Comment (mark-monteiro) [inline MediaBrowser.Api/UserLibrary/BaseItemsRequest.cs:416]: The `.` should be at the start of this line

**Pull request merged**